core/src/main/scala/kafka/server/BrokerServer.scala

```diff
@@ -153,6 +150,8 @@ class BrokerServer(
 
   var metadataPublisher: BrokerMetadataPublisher = null
 
+  var brokerFeatures: BrokerFeatures = null
```

Comment:
Hmm, since this corresponds to the features supported by the broker, it seems like this should be known statically at compile time. So I don't see an advantage to initializing it dynamically.

If we want to support simulating old brokers in junit tests, we should add this as a constructor parameter (possibly one which defaults to the latest versions of everything if you want to get fancy....)